### PR TITLE
Refine `event_buffer/` module

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

Refinement of module `event_buffer/`. Mainly on non-functional code refinement (code style, class design) and a little bit functional modification (bug fix).

A brief highlight of this PR:
- Re-design the inheritance relationship of `AtomEvent` / `CascadeEvent`.
- Change misleading function names, for example: `EventLinkedList.pop()` => `EventLinkedList.front()` (since it does not delete elements).
- Remove unused functions & variables.
- Re-organize some major functions' logic. For example: `EventLinkedList.front()`
- Fixing the bug in `EventLinkedList.__iter__()`
- Add two more test cases.

## Type of Change

- [x] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [x] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [x] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [x] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [x] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [x] Add/update the related comments
- [x] Add/update the related tests
- [x] Add/update the related documentations
- [x] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/event_buffer/event_pool.py [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
 
-from .atom_event import AtomEvent
-from .cascade_event import CascadeEvent
+from .event import ActualEvent, AtomEvent, CascadeEvent
 from .event_linked_list import EventLinkedList
 from .event_state import EventState
-from .typings import Event, EventList, List, Union
+
+
+def _pop(cntr: List[ActualEvent], event_cls_type: type) -> Union[AtomEvent, CascadeEvent]:
```
Comment (Jinyu-W) [inline maro/event_buffer/event_pool.py:11]: why not `ActualEvent` instead of `Union[..., ...]`?

File: maro/event_buffer/event_pool.py [hunk 2]
```diff
@@ -36,50 +50,34 @@ def gen(self, tick: int, event_type: object, payload: object, is_cascade: bool =
...
-            if event is not None:
-                self._append(event)
+            self._append(event)
 
-    def _append(self, event: Event):
+    def _append(self, event: ActualEvent) -> None:
         """Append event to related pool"""
-        if event:
-            # Deattach the payload before recycle.
+        if isinstance(event, CascadeEvent) or isinstance(event, AtomEvent):
```
Comment (Jinyu-W) [inline maro/event_buffer/event_pool.py:74]: Why not `isinstance(ActualEvent)`?

File: maro/event_buffer/event_pool.py [hunk 3]
```diff
@@ -36,50 +50,34 @@ def gen(self, tick: int, event_type: object, payload: object, is_cascade: bool =
...
+            # Detach the payload before recycle.
             event.payload = None
-            event._next_event_ = None
+            event.next_event = None
             event.state = EventState.FINISHED
```
Comment (Jinyu-W) [inline maro/event_buffer/event_pool.py:78]: use `EventState.Recycling` here?

File: maro/event_buffer/event_linked_list.py [hunk 4]
```diff
@@ -32,140 +31,118 @@ class EventLinkedList:
...
+        event = self._head.next_event
+        decision_events = []
+        while event is not None and event.event_type == MaroEvents.PENDING_DECISION:
+            assert isinstance(event, ActualEvent)
+            decision_events.append(event)
+            event = event.next_event
+        return decision_events
 
-    def pop(self) -> Union[Event, EventList]:
+    def front(self) -> Union[None, ActualEvent, List[ActualEvent]]:
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/event_buffer/event_linked_list.py:114]: Cool, let's call it `clear_finished_and_get_front`

**Review by Jinyu-W — DISMISSED** [review]

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**